mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala

```diff
@@ -0,0 +1,326 @@
+   * Set the solver algorithm used for optimization.
+   * Default is adamW.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setSolver(value: String): this.type = set(solver, value)
+  setDefault(solver -> AdamW)
+
+  override protected def train(dataset: Dataset[_]): FMClassifierModel = instrumented { instr =>
+    val data: RDD[(Double, OldVector)] =
+      dataset.select(col($(labelCol)), col($(featuresCol))).rdd.map {
+        case Row(label: Double, features: Vector) =>
+          require(label == 0 || label == 1, s"FMClassifier was given" +
+            s" dataset with invalid label $label.  Labels must be in {0,1}; note that" +
+            s" FMClassifier currently only supports binary classification.")
+          (label, features)
+      }
+    data.persist(StorageLevel.MEMORY_AND_DISK)
+
+    val numClasses = 2
+    if (isDefined(thresholds)) {
+      require($(thresholds).length == numClasses, this.getClass.getSimpleName +
+        ".train() called with non-matching numClasses and thresholds.length." +
+        s" numClasses=$numClasses, but thresholds has length ${$(thresholds).length}")
+    }
+
+    instr.logPipelineStage(this)
+    instr.logDataset(dataset)
+    instr.logParams(this, numFactors, fitBias, fitLinear, regParam,
+      miniBatchFraction, initStd, maxIter, stepSize, tol, solver)
+    instr.logNumClasses(numClasses)
+
+    val numFeatures = data.first()._2.size
+    instr.logNumFeatures(numFeatures)
+
+    val coefficients = _train(data, numFeatures, LogisticLoss)
+
+    val (bias, linearVector, factorMatrix) = splitCoefficients(
+      coefficients, numFeatures, $(numFactors), $(fitBias), $(fitLinear))
+
+    val model = copyValues(new FMClassifierModel(uid,
+      bias, linearVector, factorMatrix, numFeatures, numClasses))
+    model
+  }
+
+  @Since("3.0.0")
+  override def copy(extra: ParamMap): FMClassifier = defaultCopy(extra)
+}
+
+@Since("3.0.0")
+object FMClassifier extends DefaultParamsReadable[FMClassifier] {
+
+  @Since("3.0.0")
+  override def load(path: String): FMClassifier = super.load(path)
+}
+
+/**
+ * Model produced by [[FMClassifier]]
+ */
+@Since("3.0.0")
+class FMClassifierModel (
+  @Since("3.0.0") override val uid: String,
+  @Since("3.0.0") val bias: Double,
+  @Since("3.0.0") val linearVector: Vector,
+  @Since("3.0.0") val factorMatrix: Matrix,
+  @Since("3.0.0") override val numFeatures: Int,
+  @Since("3.0.0") override val numClasses: Int)
+  extends ProbabilisticClassificationModel[Vector, FMClassifierModel]
+    with FMClassifierParams with MLWritable {
+
+  @transient private lazy val oldCoefficients: OldVector =
+    combineCoefficients(bias, linearVector, factorMatrix, $(fitBias), $(fitLinear))
+
+  @transient private lazy val gradient = parseLoss(
+    LogisticLoss, $(numFactors), $(fitBias), $(fitLinear), numFeatures)
+
+  override protected def predictRaw(features: Vector): Vector = {
+    val rawPrediction: Double = gradient.getRawPrediction(features, oldCoefficients)
```

Comment:
I'd perfer to define another prediction method using `bias`/`weights`/`factors` directly, to avoid  conversions to `mlllib.vector`

Reply:
I guess then we do not need `oldCoefficients`/`gradient`

Reply:
> I guess then we do not need `oldCoefficients`/`gradient`

emm...But then I have to implement the predict code twice...Actually I want to abstract it into one class.